dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java

```diff
@@ -241,13 +257,29 @@ public Grid getAggregatedDataValues( DataQueryParams params )
 
         if ( dhisConfig.isAnalyticsCacheEnabled() )
         {
-            final DataQueryParams query = DataQueryParams.newBuilder( params ).build();
-            return queryCache.get( params.getKey(), key -> getAggregatedDataValueGridInternal( query ) ).orElseGet( ListGrid::new );
+            return getCachedOrFetch( params );
         }
 
         return getAggregatedDataValueGridInternal( params );
     }
 
+    private Grid getCachedOrFetch( final DataQueryParams params )
```

Comment:
Just a thought: Could we hide this get-cached-or-else-fetch-and-cache pattern below the cache abstraction instead of exposing it in the caller? 

I.e. could we expose a method:

`Optional<V> get( String key, Function<String, V> mappingFunction );`

in the AnalyticsCache(Wrapper) and use that?

Reply:
done